File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <bits/stdc++.h>
#define loop(i, s, n) for(int i = s; i <= n; i++)

using namespace std;

int visited[51];
int arr[52];
int res;

struct p{
    string s; int cnt;
};

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    bool exist = 0;
    for(auto t: words){
        if(t == target) exist = 1;
    }
    if(!exist) return 0;
    queue<p> q;
    q.push({begin, 0});
    visited[0] = 1;
    while(!q.empty()){
        p t = q.front(); q.pop();
        if(t.s == target) return t.cnt;
        loop(i, 0, words.size() - 1){
            int cnt = 0;
            if(visited[i + 1]) continue;
            loop(j, 0, words[i].size() - 1){
                if(words[i][j] != t.s[j]) cnt++;
                if(cnt == 2) break;
            }
            if(cnt == 1){
                q.push({words[i], t.cnt + 1});
                visited[i + 1] = 1;
            }
        }
    }
    return 0;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class P{
    String word;
    int cnt;
    P(String tword, int tcnt){
        word = tword;
        cnt = tcnt;
    }
}

class Solution {
    public int solution(String begin, String target, String[] words) {
        int[] visited = new int[words.length];
        boolean flag = false;
        for(var t: words){
            if(t.equals(target)) flag = true;
        }
        if(flag == false) return 0;
        return bfs(begin, target, words, visited);
    }
    
    int bfs(String word, String target, String[] words, int[] visited){
        Queue<P> q = new LinkedList<>();
        q.offer(new P(word, 0));
        while(!q.isEmpty()){
            P t = q.poll();
            if(t.word.equals(target)) return t.cnt;
            int rcnt = 0;
            for(int i = 0; i < words.length; i++){
                int cnt = 0;
                for(int j = 0; j < words[i].length(); j++){
                    if(t.word.charAt(j) != words[i].charAt(j)){
                        cnt++;
                    }
                }
                if(cnt <= 1){
                    q.offer(new P(words[i], t.cnt + 1));
                    visited[i] = 1;
                }
            }
        }
        return 0;
    }
}